qutebrowser/browser/shared.py

```diff
@@ -339,3 +343,37 @@ def netrc_authentication(url, authenticator):
     authenticator.setPassword(password)
 
     return True
+
+
+def choose_file(multiple: bool) -> List[str]:
+    """Select file(s) for uploading, using external command defined in config.
+
+    Args:
+        multiple: Should selecting multiple files be allowed.
+
+    Return:
+        A list of selected file paths, or empty list if no file is selected.
+        If multiple is False, the return value will have at most 1 item.
+    """
+    encoding = sys.getfilesystemencoding()
+    with tempfile.NamedTemporaryFile(mode='w+', encoding=encoding) as tmpfile:
```

Comment:
Well, looks like this indeed doesn't work out on Windows:

```
  18:59:41.412 ERROR    procs      guiprocess:_on_finished:125 Process stderr:
  Traceback (most recent call last):
  
    File "C:\Users\runneradmin\AppData\Local\Temp\pytest-of-runneradmin\pytest-0\test_select_two_files_with_sin0\py_proc.py", line 4, in <module>
  
      with open(tmp_file, 'w') as f:
  
  PermissionError: [Errno 13] Permission denied: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpsguwnqtk'
```

I should've known... see #1789 and #1927 :laughing:

So I'd suggest something like here instead:

https://github.com/qutebrowser/qutebrowser/blob/a17a2ad1738c2483bed6c24a4a15fcb8dfaf49f8/qutebrowser/commands/userscripts.py#L339-L345

(perhaps also add `prefix='qutebrowser-fileselect-'` - I wonder why we don't have that for userscripts actually)

But you'll also need to take care that the file is actually deleted, even if an exception happens or we return early. The nicest way is probably a custom context manager:

```python3
import contextlib

@contextlib.contextmanager
def cleanup_file(filename):
    try:
        yield filename
    finally:
        try:
            os.remove(filename)
        except OSError as e:
            message.error(f"Failed to delete tempfile {filename} ({e})!")
```

(untested)

Which you can then use like:

```python3
with cleanup_file(filename):
    # ...
```

Would be nice to have a unit-test or two for the context manager as well, but I can take care of that if you prefer.

Reply:
Sounds good :+1:! Where would you prefer the context manager (+tests) to live?

Reply:
I'd say probably in `utils/utils.py`.

Reply:
A question regarding windows btw, in the `.feature`-file we specify the path to the file to select using `/` but this is I guess also an issue in windows? One way would be to in the python test script do something like
```python
os.path.join(*file.split('/'))
```
but maybe there is a nicer way to handle that.

Reply:
Nope - they actually get served via a HTTP server started on localhost (`tests/end2end/fixtures/webserver_sub.py`), so something like "I open data/fileselect.html" actually turns into a URL like http://localhost:1234/data/fileselect.html

Reply:
I added the `cleanup_file` context and three tests. Two questions though:

1. If the file does not exist, should the context log an error? Alternatively we can check if it actually exists before trying to delete it.
2. Can the error in case of failure be done with `log.misc.error`? I tried to use `message.error` but then got a circular import.

Reply:
1. I think that'd be covered by the `except OSError:` already, as `FileNotFoundError` is a subclass of it
2. Oh, that's unfortunate. Yeah, feel free to use that instead, I guess it doesn't happen often and it's not something very relevant for the user anyways.

Reply:
Re 1. It is indeed covered, but maybe an error should not be logged in that case? It might be confusing if it says that a temporary file could not be removed when in fact is does not exist. What do you think?